featuretools/synthesis/deep_feature_synthesis.py

```diff
@@ -516,15 +516,28 @@ def _build_transform_features(self, all_features, entity, max_depth=0,
             input_types = groupby_prim.input_types[:]
             # if multiple input_types, only use first one for DFS
             if type(input_types[0]) == list:
-                input_types = input_types[0]
-            input_types.append(Id)
-
+                input_types = input_types[0][:]
+            id_is_input = Id in input_types
+            # Add Id if not present for the matching function
+            if not id_is_input:
+                input_types.append(Id)
+            id_index = input_types.index(Id)
```

Comment:
I think instead of checking matching inputs to get `Id` features we can use `self._features_by_type`